File: solutions/cpp/allergies/2/allergies.cpp
#include "allergies.h"

namespace allergies {

static const std::map<std::string_view, unsigned int> allergies_ = {
    {"eggs", 1},      {"peanuts", 2},    {"shellfish", 4}, {"strawberries", 8},
    {"tomatoes", 16}, {"chocolate", 32}, {"pollen", 64},   {"cats", 128}};

allergy_test::allergy_test(unsigned int score) : score_(score) {}

bool allergy_test::is_allergic_to(std::string_view allergen) const {
  const auto it = allergies_.find(allergen);
  return (it != allergies_.end()) && (score_ & it->second);
}

std::unordered_set<std::string> allergy_test::get_allergies() const {
  std::unordered_set<std::string> allergies;
  for (auto [allergy, score] : allergies_)
    if (score_ & score)
      allergies.insert(std::string(allergy));
  return allergies;
}

} // namespace allergies

File: solutions/swift/allergies/1/Sources/Allergies/Allergies.swift
struct Allergen: OptionSet {
  let rawValue: Int
  static let eggs = Allergen(rawValue: 1)
  static let peanuts = Allergen(rawValue: 2)
  static let shellfish = Allergen(rawValue: 4)
  static let strawberries = Allergen(rawValue: 8)
  static let tomatoes = Allergen(rawValue: 16)
  static let chocolate = Allergen(rawValue: 32)
  static let pollen = Allergen(rawValue: 64)
  static let cats = Allergen(rawValue: 128)
}

class Allergies {
  let allergen: Allergen
  init(_ score: Int) {
    self.allergen = Allergen(rawValue: score)
  }
  func allergicTo(item: String) -> Bool {
    switch item {
    case "eggs": return allergen.contains(.eggs)
    case "peanuts": return allergen.contains(.peanuts)
    case "shellfish": return allergen.contains(.shellfish)
    case "strawberries": return allergen.contains(.strawberries)
    case "tomatoes": return allergen.contains(.tomatoes)
    case "chocolate": return allergen.contains(.chocolate)
    case "pollen": return allergen.contains(.pollen)
    case "cats": return allergen.contains(.cats)
    default: fatalError()
    }
  }
  func list() -> [String] {
    var retval = [String]()
    if allergen.contains(.eggs) { retval.append("eggs") }
    if allergen.contains(.peanuts) { retval.append("peanuts") }
    if allergen.contains(.shellfish) { retval.append("shellfish") }
    if allergen.contains(.strawberries) { retval.append("strawberries") }
    if allergen.contains(.tomatoes) { retval.append("tomatoes") }
    if allergen.contains(.chocolate) { retval.append("chocolate") }
    if allergen.contains(.pollen) { retval.append("pollen") }
    if allergen.contains(.cats) { retval.append("cats") }
    return retval
  }
}
